### PR TITLE
First stab through internal routing

### PR DESCRIPTION
**Pull request by foolswood**
There's tidying to be done here, and almost certainly some sharp edge cases remaining.
That said it works in the basic "playing back a thing" case.

### CODE REVIEW

File: graph/src/proc_block.c [hunk 1]
```diff
@@ -279,6 +292,199 @@ void rage_proc_block_process(
...
+static rage_StepSet ** rage_step_deps(
+        rage_ProcSteps const * steps, rage_PBConnection const * cons) {
+    rage_StepSet ** deps = calloc(steps->len, sizeof(rage_StepSet *));
+    for (uint32_t i = 0; i < steps->len; i++) {
+        deps[i] = rage_step_set_new();
+    }
+    for (; cons != NULL; cons = cons->next) {
+        if (!rage_conn_is_external(cons)) {
+            rage_StepIdx const i = rage_step_for(steps, cons->sink_harness);
+            rage_StepSet * const extended_deps = rage_step_set_add(deps[i.idx], i.step);
```
Comment (ch3pjw) [inline graph/src/proc_block.c]: The set of dependencies should have the source's step, not the sink's.

File: graph/src/proc_block.c [hunk 2]
```diff
@@ -246,11 +336,349 @@ void rage_proc_block_process(
...
+        rage_Harness * sink, uint32_t sink_idx) {
+    rage_PBConnection new = {
+        .source_harness = source, .source_idx = source_idx,
+        .sink_harness = sink, .sink_idx = sink_idx,
+        .next = pb->cons
+    };
+    rage_RtBits const * old_rt = rage_rt_crit_update_start(pb->syncy);
+    rage_OrderedProcSteps ops = rage_order_proc_steps(&old_rt->steps, &new);
+    if (RAGE_FAILED(ops)) {
+        return RAGE_FAILURE_CAST(rage_Error, ops);
```
Comment (ch3pjw) [inline graph/src/proc_block.c]: finish crit update in error case

File: graph/src/proc_block.c [hunk 3]
```diff
@@ -246,11 +336,349 @@ void rage_proc_block_process(
...
+            c->next = assigned_cons;
+            os.items[i].out_buffer_allocs[c->source_idx] = c->assignment;
+            for (rage_ConnTarget * ct = c->con_tgt; ct != NULL; ct = ct->next) {
+                if (ct->tgt_harness == NULL) {  // External connection
+                    continue;  // FIXME: should the external and internal connections be mixed up?
+                }
+                uint32_t const step_idx = rage_step_idx(&os, ct->tgt_harness);
+                os.items[step_idx].in_buffer_allocs[ct->tgt_idx] = c->assignment;
+            }
+            assigned_cons = c;
```
Comment (ch3pjw) [inline graph/src/proc_block.c]: 620 and 629 probably belong together?

**Review by ch3pjw — CHANGES_REQUESTED** [review]
Some things to look at. I haven't quite got to the end of looking through `proc_block.c` as it's quite fiddly!

File: graph/src/proc_block.c [hunk 4]
```diff
@@ -246,11 +336,355 @@ void rage_proc_block_process(
...
+            }
+            if (rage_step_set_is_weak_subset(deps[i], resolved)) {
+                ordered_steps.items[resolved_idx] = steps->items[i];
+                ordered_steps.items[resolved_idx].in_buffer_allocs = calloc(
+                    steps->items[i].harness->elem->cet->inputs.len,
+                    sizeof(uint32_t));
+                ordered_steps.items[resolved_idx].out_buffer_allocs = calloc(
+                    steps->items[i].harness->elem->cet->outputs.len,
+                    sizeof(uint32_t));
+                for (
```
Comment (foolswood) [inline graph/src/proc_block.c]: rage_alloc_int_array?

**Review by ch3pjw — APPROVED** [review]
I think I kinda handwavily understand what this is doing!

**Pull request merged**